# Ejercicios Pair Programming 
## Lección 10 - EDA I

Empezamos una nueva lección, y este es el momento de poner en práctica los conocimientos adquiridos del Análisis Exploratorio de Datos. Si recordamos, en la lecciones de Pandas estuvimos usando un dataset sobre ataques de tiburones, y que hicimos algunos cambios a lo largo de los ejercicios, incluimos filas, columnas e incluso limpiamos algunas colummas 💪🏽! Para todas los ejercicios de pair programming de EDA I y II usaremos el dataframe generado en el ejercicio de Pandas VII.

En el ejercicio de hoy nos familiarizaremos con el dataset completo sobre el que trabajaremos en la parte de limpieza.

Pero para hacer una exploración buena del dataset no lo podemos hacer como pollo sin cabeza a ver que nos encontramos. Necesitamos hacernos una serie de preguntas o hipótesis que queramos contestar. En este caso os plantemos las siguientes preguntas a contestar a lo largo de los ejercicios de pair programming de este módulo:

### Hipótesis

Las hipótesis que os planteamos son:
- ¿Es Australia el sitio más peligroso y letal para estar relajada en la playa?
- ¿Hay diferencias entre los países en los ataques a hombres y mujeres?
- ¿Cuáles son las edades que más sufren ataques?
- Independientemente de la edad, ¿sufren los hombres más ataques que las mujeres?
- ¿En qué mes ocurren más ataques?
- ¿Cuál es la relación entre la especie y el tipo de ataque (si es fatal o no)?
- ¿Cómo han evolucionado los ataques a lo largo del tiempo?

In [1]:
import pandas as pd

In [2]:
pd.options.display.max_columns = None

In [3]:
df = pd.read_csv("attacks4.csv", index_col = 0)

In [4]:
df.head(2)

case_number    year        type country        area  \
0    2018.06.25  2018.0     Boating     USA  California   
1  2018.06.03.a  2018.0  Unprovoked  BRAZIL  Pernambuco   

                      location  activity                    name age  \
0  Oceanside, San Diego County  Paddling             Julie Wolfe  57   
1        Piedade Beach, Recife  Swimming  Jose Ernesto da Silva   18   

             time     species_  \
0           18h00  White shark   
1  Late afternoon  Tiger shark   

                                                href      siglo  unnamed:_0  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI           0   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI           6   

                                              injury  mes fatal sex  
0  No injury to occupant, outrigger canoe and pad...  Jun     N   F  
1                                              FATAL  Jun     Y   M

### Objetivos

Los objetivos de hoy son:
1. Haced una exploración inicial del dataframe:

- ¿Cuál es el número de filas y columnas?

In [5]:
print(f"El número de filas de nuestro DataFrame es {df.shape[0]} y el número de columnas es {df.shape[1]}")

El número de filas de nuestro DataFrame es 1672 y el número de columnas es 18


- ¿Cuál es el porcentaje de valores nulos en cada columna? Presenta los resultados en un dataframe, donde las columnas sean el nombre de las variables y el porcentaje de nulos.

In [6]:
df_porcentaje = pd.DataFrame(df.isnull().sum()*100/df.shape[0]).T
df_porcentaje

case_number  year  type   country      area  location  activity      name  \
0          0.0   0.0   0.0  0.598086  2.751196  3.050239  1.854067  0.538278   

        age      time  species_  href  siglo  unnamed:_0    injury        mes  \
0  9.210526  9.569378  7.535885   0.0    0.0         0.0  0.478469  13.038278   

      fatal       sex  
0  5.801435  0.897129

- ¿Qué tipos de datos hay en cada columna?

In [7]:
df.dtypes

case_number     object
year           float64
type            object
country         object
area            object
location        object
activity        object
name            object
age             object
time            object
species_        object
href            object
siglo           object
unnamed:_0       int64
injury          object
mes             object
fatal           object
sex             object
dtype: object

- Extraed la información general del dataframe.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1672 entries, 0 to 1671
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   case_number  1672 non-null   object 
 1   year         1672 non-null   float64
 2   type         1672 non-null   object 
 3   country      1662 non-null   object 
 4   area         1626 non-null   object 
 5   location     1621 non-null   object 
 6   activity     1641 non-null   object 
 7   name         1663 non-null   object 
 8   age          1518 non-null   object 
 9   time         1512 non-null   object 
 10  species_     1546 non-null   object 
 11  href         1672 non-null   object 
 12  siglo        1672 non-null   object 
 13  unnamed:_0   1672 non-null   int64  
 14  injury       1664 non-null   object 
 15  mes          1454 non-null   object 
 16  fatal        1575 non-null   object 
 17  sex          1657 non-null   object 
dtypes: float64(1), int64(1), object(16)
memory usage

2. Cread dos dataframes nuevos, uno que incluya solo las variables categóricas y otro que incluya solo las variables numéricas.

In [24]:
df_categoricas = pd.DataFrame(df.describe(include = "object"))
df_categoricas

case_number        type country     area  \
count         1672        1672    1662     1626   
unique        1672           6      95      236   
top     2018.06.25  Unprovoked     USA  Florida   
freq             1        1432     741      346   

                                location activity  name   age   time  \
count                               1621     1641  1663  1518   1512   
unique                              1310      404  1536    94    248   
top     New Smyrna Beach, Volusia County  Surfing  male    19  11h00   
freq                                  73      426    97    70     74   

           species_                                               href  \
count          1546                                               1672   
unique          856                                               1672   
top     White shark  http://sharkattackfile.net/spreadsheets/pdf_di...   
freq             76                                                  1   

           siglo injury   mes fatal   sex  
count       1672   1664  1454  1575  1657  
unique         2   1173    12     2     2  
top     siglo-XX  FATAL   Aug     N     M  
freq         887    154   153  1258  1435

In [10]:
df_numericas = pd.DataFrame(df.describe()).T
df_numericas

count         mean          std  min      25%     50%     75%  \
year        1672.0  1832.688397   528.969901  0.0  1964.00  1997.0  2008.0   
unnamed:_0  1672.0  2574.174043  1756.622536  0.0  1218.25  2210.0  3687.5   

               max  
year        2018.0  
unnamed:_0  6285.0

- ¿Cuáles son los principales estádisticos de los dataframes?

De las variables numéricas sólo tenemos la columna 'year', y sus principales estadísticos tienen un valor de:
- Hay un total de 1672 filas.
- La media es de 1832,68 años.
- La desviación estándar es de 528,96 con respecto a la media.
- El año menor es el 0 y el año mayor es el 2018.
- El cuartil 1 es el año 1964 (el 25 % de los datos es inferior a 1964)
- El cuartil 2 o mediana es el año 1997.
- El cuartil 3 es el año 2008.

De las variables categóricas tenemos el resto de las columnas y sus principales estadísticos son:
- Número total de filas.
- Los valores únicos de cada columna.
- Top, que es la moda, valor que más se repite.
- Y la frecuencia de dicho top.

- Extraed los valores únicos su frecuencia. BONUS presenta los resultados en un dataframe.

Donde variable es cada una de las variables categóricas que tenemos en el dataset y n_unicos es una lista con los nombres de los valores únicos de cada variable.

In [17]:
#frecuencia
for col in df_categoricas.columns:
    print("--------------")
    print(f"En la columna '{col}' tenemos los siguientes valores:")
    print(pd.DataFrame(df_categoricas[col].value_counts()))
    print("--------------")

--------------
En la columna 'count' tenemos los siguientes valores:
      count
1672      4
1662      1
1626      1
1621      1
1641      1
1663      1
1518      1
1512      1
1546      1
1664      1
1454      1
1575      1
1657      1
--------------
--------------
En la columna 'unique' tenemos los siguientes valores:
      unique
2          3
1672       2
6          1
95         1
236        1
1310       1
404        1
1536       1
94         1
248        1
856        1
1173       1
12         1
--------------
--------------
En la columna 'top' tenemos los siguientes valores:
                                                    top
2018.06.25                                            1
Unprovoked                                            1
USA                                                   1
Florida                                               1
New Smyrna Beach, Volusia County                      1
Surfing                                               1
male                  

In [19]:
df_categoricas.index

Index(['case_number', 'type', 'country', 'area', 'location', 'activity',
       'name', 'age', 'time', 'species_', 'href', 'siglo', 'injury', 'mes',
       'fatal', 'sex'],
      dtype='object')

In [28]:
for col in df_categoricas.columns:
    print("--------------")
    print(col)
    print(df[col].unique())
    print("--------------")

--------------
case_number
['2018.06.25' '2018.06.03.a' '2018.05.26.b' ... 'ND.0003' 'ND.0002'
 'ND.0001']
--------------


NameError: name 'df_bonus' is not defined

In [14]:
#df_frecuencia = pd.DataFrame()
for col in df_categoricas.columns:
    print(df_categoricas.index)
    print(df_categoricas[col].unique())
    print(df_categoricas[col].value_counts())
    print("--------------")

Index(['case_number', 'type', 'country', 'area', 'location', 'activity',
       'name', 'age', 'time', 'species_', 'href', 'siglo', 'injury', 'mes',
       'fatal', 'sex'],
      dtype='object')
[1672 1662 1626 1621 1641 1663 1518 1512 1546 1664 1454 1575 1657]
1672    4
1662    1
1626    1
1621    1
1641    1
1663    1
1518    1
1512    1
1546    1
1664    1
1454    1
1575    1
1657    1
Name: count, dtype: int64
--------------
Index(['case_number', 'type', 'country', 'area', 'location', 'activity',
       'name', 'age', 'time', 'species_', 'href', 'siglo', 'injury', 'mes',
       'fatal', 'sex'],
      dtype='object')
[1672 6 95 236 1310 404 1536 94 248 856 2 1173 12]
2       3
1672    2
6       1
95      1
236     1
1310    1
404     1
1536    1
94      1
248     1
856     1
1173    1
12      1
Name: unique, dtype: int64
--------------
Index(['case_number', 'type', 'country', 'area', 'location', 'activity',
       'name', 'age', 'time', 'species_', 'href', 'siglo', 'injury', 'mes',


### Conclusiones

De qué nos sirven estos números y dataframe? Extraed conclusiones de ellos.

Por ejemplo:
- ¿Hay alguna columna que no tenga sentido su tipo de datos?

- ¿Hay alguna columna que no tenga sentido su contenido?

- ¿Hay alguna columna que tenga demasiadas categorías o valores que nos pueda hacer difícil trabajar con ella?

Happy coding 🦈